Replicate Expiry ILM configs while site replication

shtripat opened the pull request:
## Community Contribution License
All community contributions in this pull request are licensed to the project maintainers
under the terms of the [Apache 2 license] (https://www.apache.org/licenses/LICENSE-2.0). 
By creating this pull request I represent that I have the right to license the 
contributions to the project maintainers under the Apache 2 license.

## Description


## Motivation and Context


## How to test this PR?
Refer corresponding `mc` PR https://github.com/minio/mc/pull/4698 for detailed steps

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Unit tests added/updated
- [ ] Internal documentation updated
- [ ] Create a documentation update request [here](https://github.com/minio/docs/issues/new?label=doc-change,title=Doc+Updated+Needed+For+PR+github.com%2fminio%2fminio%2fpull%2fNNNNN)

shtripat committed:
Replicate expiry lifecycle config while site replication

Signed-off-by: Shubhendu Ram Tripathi <shubhendu@minio.io>

shtripat committed:
Fixed lint issues

Signed-off-by: Shubhendu Ram Tripathi <shubhendu@minio.io>

poornas commented on cmd/bucket-lifecycle-handlers.go:
```diff
@@ -33,6 +33,8 @@ import (
 	// Lifecycle configuration file.
 	bucketLifecycleConfig = "lifecycle.xml"
+	// Expiry Lifecycle configuration file
+	expiryBucketLifecycleConfig = "expiry-lifecycle.xml"
```
may want to add a field in Lifecycle config to track last update timestamp for expiry rules. Otherwise, there will be redundant lifecycle configs and consistency problems.

shtripat replied:
Ok. Will update the PR accordingly.

poornas commented on cmd/site-replication.go:
```diff
@@ -196,23 +196,26 @@ type SiteReplicationSys struct {
+// srStateV2 represents version 2 of the site replication state persistence
 // format.
-type srStateV1 struct {
+type srStateV2 struct {
```
this can be left as V1 for now

shtripat committed:
Incorporated review comments

Signed-off-by: Shubhendu Ram Tripathi <shubhendu@minio.io>

shtripat committed:
Fix mint failures

Signed-off-by: Shubhendu Ram Tripathi <shubhendu@minio.io>

shtripat committed:
Added option to enable back ILM expiry for replication

Signed-off-by: Shubhendu Ram Tripathi <shubhendu@minio.io>

shtripat committed:
Show ILM expiry replication details in `mc admin replicate status`

Signed-off-by: Shubhendu Ram Tripathi <shubhendu@minio.io>

shtripat committed:
Support replication of deleted ILM expiry rules

Signed-off-by: Shubhendu Ram Tripathi <shubhendu@minio.io>

shtripat committed:
Added test script for ILM expiry rules replication

Signed-off-by: Shubhendu Ram Tripathi <shubhendu@minio.io>

shtripat committed:
Explicitly set `ExpiryUpdatedAt` as nil while list

Signed-off-by: Shubhendu Ram Tripathi <shubhendu@minio.io>

poornas commented on cmd/site-replication.go:
```diff
@@ -548,6 +552,7 @@ func (c *SiteReplicationSys) AddPeerClusters(ctx context.Context, psites []madmi
 		Peers:                   joinReq.Peers,
 		ServiceAccountAccessKey: svcCred.AccessKey,
+		ReplicateILMExpiry:      opts.ReplicateILMExpiry,
 	}
```
need to pass the ilm expiry setting to. SRPeerJoin

shtripat replied:
Ok

poornas commented on cmd/bucket-metadata-sys.go:
```diff
@@ -177,6 +178,40 @@ func (sys *BucketMetadataSys) save(ctx context.Context, meta BucketMetadata) err
+				return updatedAt, err
+			}
+		}
+		return sys.updateAndParse(ctx, bucket, configFile, cfgData, false)
```
why do you need this change? expiryUpdatedAt timestamp is already preserved in metadata for lifecycle changes. You will just need to return it in GetLifecycleConfig when config is nil

shtripat replied:
Ok, let me try that.

shtripat replied:
This is required because in normal delete it sets `nil` for config file. In case of lifecycle config and delete, we still want to retain the latest `ExpiryUpdatedAt` value with no rules, so that while sync it would figure out on other peers that peer with latest `ExpiryUpdatedAt` has no rules (i.e. deleted) and we need to delete as well. 
Basically for replicating the removal of expiry rules across sites.
Without this, other peers would find their `ExpiryUpdatedAt` value as bigger than the site where this delete actually happened, and they would replicate the rules back.

shtripat committed:
Replicate ILM expiry replication settings

Also:
 - fixed a panic
 - consider all expiry rules for replication

Signed-off-by: Shubhendu Ram Tripathi <shubhendu@minio.io>

shtripat committed:
Use latest minio-go changes

Signed-off-by: Shubhendu Ram Tripathi <shubhendu@minio.io>

shtripat committed:
Fix test failures

Signed-off-by: Shubhendu Ram Tripathi <shubhendu@minio.io>

poornas commented on cmd/site-replication.go:
```diff
@@ -1834,6 +1872,35 @@ func (c *SiteReplicationSys) syncToAllPeers(ctx context.Context) error {
+			var expLclCfg lifecycle.Lifecycle
+			expLclCfg.XMLName = meta.lifecycleConfig.XMLName
+			for _, rule := range meta.lifecycleConfig.Rules {
+				if !rule.Expiration.IsNull() || !rule.NoncurrentVersionExpiration.IsNull() {
```
need additional filtering of transition settings if any

shtripat replied:
Ack

poornas commented on cmd/site-replication.go:
```diff
@@ -3919,21 +4194,22 @@ func getSRBucketDeleteOp(isSiteReplicated bool) SRBucketDeleteOp {
+	})
+	if err != nil {
+		return err
+	}
```
fetching site replication status across sites for all buckets in one shot has an issue - depending on number of buckets and other metadata and number of sites in the site replication, this is not a reliable info to use in healing because one/more concurrent updates could happen in the meanwhile.   

Please leave the call in lines 3928-3931 as is and just include your ILMExpiryRules. The ShowDeleted flag is specific to bucket healing to propagate bucket deletes and shouldn't be dropped.

shtripat replied:
Ack

poornas commented on cmd/site-replication.go:
```diff
@@ -5516,3 +5872,79 @@ func (c *SiteReplicationSys) getSiteMetrics(ctx context.Context) (madmin.SRMetri
+		XMLName:         xmlName,
+		Rules:           rules,
+		ExpiryUpdatedAt: &updatedAt,
+	}
```
validate the finalLcCfg before marshalling

shtripat replied:
Ack

shtripat committed:
Filter transition rules and replicate others

Also replicate expiry rules if all sites have flag enabled

Signed-off-by: Shubhendu Ram Tripathi <shubhendu@minio.io>

shtripat committed:
Enable healing of site replication state

While few remote sites are down, any changes to the site
replication state e.g. ReplicateILMExpiry flag should get
replicated to site when it's back online. This would be done
exactly before healing any ILM rules changes during remote
sites were down, so that these flags are taken in consideration.

Signed-off-by: Shubhendu Ram Tripathi <shubhendu@minio.io>

poornas commented on cmd/site-replication.go:
```diff
@@ -3918,21 +4231,46 @@ func getSRBucketDeleteOp(isSiteReplicated bool) SRBucketDeleteOp {
+				fmt.Errorf("Unable to heal site replication state for peer %s from peer %s : %w",
+					pi.Name, c.state.Name, err)))
+		}
+	}
```
you can't blindly copy all settings over - sync/async mode and bandwidth throttling are meant to be unidirectional. Making a change on one site should not be copied over to others. Ensure that healing happens only on the site with the latest update to the remaining on just the ilm setting

Reviewed more , once above issue is fixed this should be good to merge. Nice work!

shtripat replied:
Sure @poornas , at receiving end I will make sure only `ReplicateILMExpiry` flag is replicated and `UpdatedAt` time get updated. Will change and update.

shtripat committed:
Heal only valid updated peer states

Signed-off-by: Shubhendu Ram Tripathi <shubhendu@minio.io>

poornas commented on cmd/site-replication.go:
```diff
@@ -3918,21 +4251,81 @@ func getSRBucketDeleteOp(isSiteReplicated bool) SRBucketDeleteOp {
+
+	for dID, ps := range info.PeerStates {
+		// If latest peers ILM expiry flags are equal to current peer, no need to heal
+		flagEual := true
```
```suggestion
		flagEqual := true
```

poornas commented on cmd/site-replication.go:
```diff
@@ -3918,21 +4251,81 @@ func getSRBucketDeleteOp(isSiteReplicated bool) SRBucketDeleteOp {
+				break
+			}
+		}
+		if flagEual {
```
```suggestion
		if flagEqual {
```

poornas commented on cmd/site-replication.go:
```diff
@@ -3918,21 +4251,81 @@ func getSRBucketDeleteOp(isSiteReplicated bool) SRBucketDeleteOp {
+		flagEual := true
+		for id, peer := range latestPeers {
+			if !(ps.Peers[id].ReplicateILMExpiry == peer.ReplicateILMExpiry) {
+				flagEual = false
```
```suggestion
				flagEqual = false
```

poornas commented on cmd/bucket-lifecycle-handlers.go:
```diff
@@ -86,6 +91,11 @@ func (api objectAPIHandlers) PutBucketLifecycleHandler(w http.ResponseWriter, r
 		return
 	}
 
+	if bucketLifecycle.HasExpiry() || *lcCfg.ExpiryRuleRemoved {
```
```suggestion
	if bucketLifecycle.HasExpiry() || (lcCfg != nil && *lcCfg.ExpiryRuleRemoved) {
```

shtripat committed:
Apply ilm expiry replication changes to all sites

Signed-off-by: Shubhendu Ram Tripathi <shubhendu@minio.io>

vadmeste commented on cmd/admin-handlers-site-replication.go:
```diff
@@ -334,6 +343,7 @@ func (a adminAPIHandlers) SiteReplicationStatus(w http.ResponseWriter, r *http.R
 		opts.Users = true
 		opts.Policies = true
 		opts.Groups = true
+		opts.ILMExpiryRules = true
```
wouldn't ilm healing replication disabled by default ?

shtripat replied:
you mean, if ILM expiry replication not enable, don't report their replication status in `mc admin replicate status ALIAS`?  Currently it would show status as below if ILM expiry replication is disabled
```
ILM Expiry Rules replication status:
●  0/1 ILM Expiry Rules in sync
```

vadmeste commented on cmd/admin-handlers-site-replication.go:
```diff
@@ -422,12 +441,37 @@ func (a adminAPIHandlers) SRPeerEdit(w http.ResponseWriter, r *http.Request) {
+func (a adminAPIHandlers) SRStateEdit(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+
+	objectAPI, _ := validateAdminReq(ctx, w, r, policy.SiteReplicationAddAction)
```
It looks to me that the action should be 'policy.SiteReplicationOperationAction'

shtripat replied:
Ah yes, thats a copy-paste issue. Will correct.

poornas commented on cmd/bucket-lifecycle-handlers.go:
```diff
@@ -86,6 +91,11 @@ func (api objectAPIHandlers) PutBucketLifecycleHandler(w http.ResponseWriter, r
 		return
 	}
 
+	if bucketLifecycle.HasExpiry() || (lcCfg != nil && *lcCfg.ExpiryRuleRemoved) {
```
```suggestion
	if bucketLifecycle.HasExpiry() || (lcCfg != nil && lcCfg.ExpiryRuleRemoved != nil && *lcCfg.ExpiryRuleRemoved) {
```
ExpiryRuleRemoved is also a pointer - can you rename this to ExpiryRuleRemovedPtr

vadmeste commented on cmd/site-replication.go:
```diff
@@ -1681,6 +1721,35 @@ func (c *SiteReplicationSys) PeerBucketQuotaConfigHandler(ctx context.Context, b
 }
 
+// PeerBucketLCConfigHandler - copies/deletes lifecycle config to local cluster
+func (c *SiteReplicationSys) PeerBucketLCConfigHandler(ctx context.Context, bucket string, expLCConfig *string, updatedAt time.Time) error {
```
Any strong reason why this is '*string', it looks []byte is enough

shtripat replied:
No specific reason. Just to keep inline with other XML data like `ObjectLockConfig` and `SSEConfig` under `SRBucketMeta` struct `ExpiryLCConfig` also kept as `*string` and the same passed to ` globalSiteReplicationSys.PeerBucketLCConfigHandler`

vadmeste commented on cmd/site-replication.go:
```diff
@@ -5510,3 +6061,100 @@ func (c *SiteReplicationSys) getSiteMetrics(ctx context.Context) (madmin.SRMetri
 }
+
+// mergeWithCurrentLCConfig - merges the given ilm expiry configuration with existing for the current site and returns
+func mergeWithCurrentLCConfig(ctx context.Context, bucket string, expLCCfg *string, updatedAt time.Time) ([]byte, error) {
```
Is merge logic really needed ? We already require the newly added site replication to be empty, so I guess it is enough to overwrite the ILM policy.

Let's also forbid adding a new ILM transition rule when the site replication is enabled

shtripat replied:
This merge is need for the below scenario.

Say we have 1 transition + 2 expiry rules under stite1 and then we enable site replication to site2 with `--replicate-ilm-expiry` flag. Now only the two expiry rules from site1 would be replicated to site2.

Later say we update contents of the expiry rules in site2. As the latest update time is higher on site2, these rules need to be replicated back to site1 as part of healing. Now while writing these expiry rules from site2 to site1, we should not overwrite all rules of site1 but merge with exiting ones on site1.

Without this merge, all transition rules on site1were getting lost as part of heal action.
As part of merge first we find all local rules, if incoming expiry rules already there in local list we update their content else just add to the local list. Then finally this list is returned for the local site for write.

There is no adding of any new ILM transition rule when site replication enabled.

vadmeste reviewed (commented):
Few comments from first round of review

shtripat committed:
Corrected policy action

Signed-off-by: Shubhendu Ram Tripathi <shubhendu@minio.io>

shtripat committed:
Remove only expiry rules while heal (if removed on remote)

Signed-off-by: Shubhendu Ram Tripathi <shubhendu@minio.io>

shtripat committed:
Copy anything but transition rules while replication

This would enable copying of tags, prefixes etc while replication.
Also enabled to figure out if removal of ILM expiry rules needs
replication. This check would be performed at server side now, instead
of expecting client to send a flag.

Signed-off-by: Shubhendu Ram Tripathi <shubhendu@minio.io>

shtripat committed:
Added tests for three site replication

Signed-off-by: Shubhendu Ram Tripathi <shubhendu@minio.io>

poornas commented on cmd/bucket-lifecycle-handlers.go:
```diff
@@ -86,6 +87,39 @@ func (api objectAPIHandlers) PutBucketLifecycleHandler(w http.ResponseWriter, r
+			// set the flag that expiry rule removed as part of request
+			if _, ok := updatedRules[rl.ID]; ok && !rl.Expiration.IsNull() || !rl.NoncurrentVersionExpiration.IsNull() {
+				expiryRuleRemoved = true
+			}
```
this check is flawed - if a rule had just transition flags and is unchanged, expiryRuleRemoved would be true.

```
updtdRule, ok := updatedRules[rl.ID]
// original rule had expiry that is no longer in the new config, or rule is present but missing expiration flags
if (!rl.Expiration.IsNull() || !rl.NoncurrentVersionExpiration.IsNull()) && 
   (!ok || (updtdRule.Expiration.IsNull() && updtdRule.NoncurrentVersionExpiration.IsNull()){
				expiryRuleRemoved = true
}
```

shtripat replied:
Ah yes. didn't realize that. Will update. Thanks!

shtripat committed:
Corrected check for `expiryRuleRemoved`as suggested

Signed-off-by: Shubhendu Ram Tripathi <shubhendu@minio.io>

shtripat committed:
Heal everything but tranisition details of rules

Signed-off-by: Shubhendu Ram Tripathi <shubhendu@minio.io>

shtripat committed:
Correct nil check

Signed-off-by: Shubhendu Ram Tripathi <shubhendu@minio.io>

poornas reviewed (approved):
LGTM

harshavardhana commented on cmd/admin-handlers-site-replication.go:
```diff
@@ -68,6 +69,12 @@ func (a adminAPIHandlers) SiteReplicationAdd(w http.ResponseWriter, r *http.Requ
 }
 
+func getSRAddOptions(r *http.Request) (opts madmin.SRAddOptions) {
+	q := r.Form
+	opts.ReplicateILMExpiry = q.Get("replicateILMExpiry") == "true"
```
```suggestion
	opts.ReplicateILMExpiry = r.Form.Get("replicateILMExpiry") == "true"
```

harshavardhana commented on cmd/admin-handlers-site-replication.go:
```diff
@@ -398,6 +410,13 @@ func (a adminAPIHandlers) SiteReplicationEdit(w http.ResponseWriter, r *http.Req
 }
 
+func getSREditOptions(r *http.Request) (opts madmin.SREditOptions) {
+	q := r.Form
+	opts.DisableILMExpiryReplication = q.Get("disableILMExpiryReplication") == "true"
+	opts.EnableILMExpiryReplication = q.Get("enableILMExpiryReplication") == "true"
+	return
```
```suggestion
	opts.DisableILMExpiryReplication = r.Form.Get("disableILMExpiryReplication") == "true"
	opts.EnableILMExpiryReplication = r.Form.Get("enableILMExpiryReplication") == "true"
	return
```

shtripat replied:
Sent a new PR https://github.com/minio/minio/pull/18499 for the same

harshavardhana reviewed (approved):
Just that we don't need to do q := r.Form - a shallow copy is not necessary.
